src/lib/peer_trust/peer_trust.ml

```diff
@@ -46,15 +47,216 @@ end) (Now : sig
+            verb peer action simple_new.trust ;
+          Deferred.unit
+    in
+    Db.set db ~key:peer ~data:new_record
 end
+
+let%test_module "peer_trust" =
+  ( module struct
+    (* Mock the current time *)
+    module Mock_now = struct
+      let current_time = ref Time.epoch
+
+      let now () = !current_time
+
+      let advance span = current_time := Time.add !current_time span
+    end
+
+    module Mock_record = Record.Make (Mock_now)
+    module Db = Key_value_database.Make_mock (Int) (Record)
+
+    module Action = struct
+      type t = Insta_ban | Slow_punish | Slow_credit | Big_credit
+      [@@deriving sexp]
+
+      let to_trust_response t =
+        match t with
+        | Insta_ban -> Trust_response.Insta_ban
+        | Slow_punish -> Trust_response.Trust_decrease (max_rate 1.)
+        | Slow_credit -> Trust_response.Trust_increase (max_rate 1.)
+        | Big_credit -> Trust_response.Trust_increase 0.2
+    end
+
+    module Peer_trust_test = Make (Int) (Mock_now) (Action) (Db)
+
+    (* We want to check the output of the pipe in these tests, but it's
+       synchronous, so we need to read from it in a different "thread",
+       otherwise it would hang. *)
+    let ban_pipe_out = ref []
+
+    let assert_ban_pipe expected =
+      [%test_eq: int list] expected !ban_pipe_out ;
+      ban_pipe_out := []
+
+    let setup_mock_db () =
+      let res = Peer_trust_test.create ~db_dir:"fake" in
+      don't_wait_for
+      @@ Strict_pipe.Reader.iter_without_pushback res.bans_reader ~f:(fun v ->
+             ban_pipe_out := v :: !ban_pipe_out ) ;
+      res
+
+    let nolog = Logger.null ()
+
+    let%test "Peers are unbanned and have 0 trust at initialization" =
+      let db = setup_mock_db () in
+      match Peer_trust_test.lookup db 0 with
+      | {trust= 0.0; banned= Unbanned} -> assert_ban_pipe [] ; true
+      | _ -> false
+
+    let%test "Insta-bans actually do so" =
+      let db = setup_mock_db () in
+      Thread_safe.block_on_async_exn (fun () ->
+          Peer_trust_test.record db nolog 0 Insta_ban ) ;
+      match Peer_trust_test.lookup db 0 with
+      | {trust= -1.0; banned= Banned_until time} ->
+          [%test_eq: Time.t] time
+          @@ Time.add !Mock_now.current_time Time.Span.day ;
+          assert_ban_pipe [0] ;
+          true
+      | _ -> false
+
+    let%test "trust decays by half in 24 hours" =
+      let db = setup_mock_db () in
+      Thread_safe.block_on_async_exn (fun () ->
+          Peer_trust_test.record db nolog 0 Action.Big_credit ) ;
+      match Peer_trust_test.lookup db 0 with
+      | {trust= start_trust; banned= Unbanned} -> (
+          Mock_now.advance Time.Span.day ;
+          assert_ban_pipe [] ;
+          match Peer_trust_test.lookup db 0 with
+          | {trust= decayed_trust; banned= Unbanned} ->
+              decayed_trust =. start_trust /. 2.0
+          | _ -> false )
+      | _ -> false
+
+    let do_constant_rate rate f =
```

Comment:
should the name of this function have `week` in it?

Reply:
the rate is number of times per second? maybe mention in a comment

Reply:
Specified the rate is per week in a comment. I don't want to make the function name even longer, and the duration is in a comment